[Vue] Add "Expand v-model" intention

panstromek opened the pull request:
![vmodelexpandfixed](https://user-images.githubusercontent.com/25121817/62003713-7cf25380-b11b-11e9-8528-9788d3189c33.gif)

First time trying to contribute to IntelliJ plugin, so I chose a little intention I want to use sometimes. Let me know how should I polish this ;), it's probably not aligned with all conventions.

I also don't know if 192 is the right target branch - it seems that it is the most active branch for Vue plugin, so I forked that instead of master.
 
I already signed Contributors agreement.

piotrtomiak commented on vuejs/src/org/jetbrains/vuejs/intentions/expandVModel/VueExpandVModelIntention.kt:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    private fun isValid(element: PsiElement): Boolean {
+        val text = element.text
```
I think you could use `VueAttributeNameParser` here to extract information about the attribute. It parses modifiers and arguments. Please use `XmlAttribute#getName` instead of `getText` to get attribute name

piotrtomiak commented on vuejs/src/org/jetbrains/vuejs/intentions/expandVModel/VueExpandVModelIntention.kt:
```diff
@@ -0,0 +1,70 @@
+        var assignedValue = if (isComponent) "\$event" else "\$event.target.value"
+        var eventName = "@input"
+        var nextNeedsParens = false
+        getModifiers(directiveText).forEach {
```
Please use `AttributeNameParser` for extracting modifiers. Does the order of modifiers matter on `v-model`?

panstromek replied:
yes, order does matter (even though in this case not too much)

panstromek replied:
.. Actually - after some investigation I found out it doesn't - and code generated by vue-compiler is a bit different for some cases, too. I will need to update the code.

panstromek replied:
(order does matter only for events)

panstromek replied:
Refactored it to not care about order.

piotrtomiak commented on vuejs/src/org/jetbrains/vuejs/intentions/expandVModel/VueExpandVModelIntention.kt:
```diff
@@ -0,0 +1,70 @@
+    override fun isAvailable(project: Project, editor: Editor?, element: PsiElement): Boolean {
+        return element.node.elementType == XmlElementType.XML_NAME &&
+                element.parent?.node?.elementType == XmlElementType.XML_ATTRIBUTE &&
+                isValid(element)
```
Please add a check whether you are in a Vue context - `org.jetbrains.vuejs.index.isVueContext(PsiElement)`

panstromek replied:
I thought about that, too. Will do ;)

piotrtomiak commented on vuejs/src/org/jetbrains/vuejs/intentions/expandVModel/VueExpandVModelIntention.kt:
```diff
@@ -0,0 +1,70 @@
+        val directiveText = parent.name
+        val isComponent = parentTag.descriptor is VueElementDescriptor
+        var assignedValue = if (isComponent) "\$event" else "\$event.target.value"
+        var eventName = "@input"
```
I think the event name should be taken from the component's `model` property:
```
(parentTag.descriptor as VueElementDescriptor).getModel().event
```
You need to add the `getModel()` function similarly to `getProps()`, but I guess you have to take into account only the first source.

panstromek replied:
` I guess you have to take into account only the first source.`  - what do you mean by that? There can be only one v-model on a component - unless there's a syntax error. In that case, User will still want to transform the model that is under the caret, which is not necessarily the one returned from the descriptor..

piotrtomiak replied:
I meant the implementation for `getModel` function in `VueElementDescriptor`; it has `sources` field, which usually has only one element, but sometimes more. This ambiguity is present, because in same rare cases we are not sure which of the candidates is really used in the template.

panstromek replied:
I misunderstood the meaning of VueElementDescriptor at first. I implemented the method. Used first source but still left defaults (`@input`, `:value`). Those are 90% times the correct ones.

piotrtomiak commented on vuejs/src/org/jetbrains/vuejs/intentions/expandVModel/VueExpandVModelIntention.kt:
```diff
@@ -0,0 +1,70 @@
+        val modelVariableName = parentAttr.value
+        CommandProcessor.getInstance().executeCommand(project, {
+            WriteAction.run<RuntimeException> {
+                parentAttr.name = ":value"
```
Same for `value` name, please take it from the component's `model` property.

panstromek replied:
done ;)

panstromek commented:
@piotrtomiak Thanks for review, I will look at it ;)

One more problem I have is that project fails to compile because of missing or private methods in JS plugin - I suppose there is some newer version of it internally in JetBrains.. I had to comment out or inline that code to make it build and run (it was unrelated to this PR). Can I do something about it?

![image](https://user-images.githubusercontent.com/25121817/62067615-6b509f00-b234-11e9-95da-6914af017212.png)

I had similar problems before. Dennis Ushakov told be that tests are disabled because there are some unpublished changes - but that was few months ago already ( here: https://youtrack.jetbrains.com/issue/WEB-37226 ).

I didn't even try to run tests this time, I'll try to add some and run.

piotrtomiak commented:
@panstromek This compilation problem is present, because you are on `192` branch, which is already ahead of the latest published build - 2019.2 . It seems that until the first 2019.2.1 EAP is published you should actually checkout `192.5728` branch. You can create PR for that branch as well. I am sorry for the inconvenience.

piotrtomiak commented on vuejs/src/org/jetbrains/vuejs/intentions/expandVModel/VueExpandVModelIntention.kt:
```diff
@@ -16,7 +16,8 @@ class VueExpandVModelIntention : JavaScriptIntention() {
 
     override fun isAvailable(project: Project, editor: Editor?, element: PsiElement): Boolean {
-        return element.node.elementType == XmlElementType.XML_NAME &&
+        return org.jetbrains.vuejs.index.isVueContext(element) &&
```
The vue context check might be an expensive operation as it needs to hit the cache and possibly filesystem. The `isAvailable` method is run on every PsiElement in the file, so it is better to move the context either just before `isValid` call or maybe better after. I think it would be good to add an import for `isVueContext`.

panstromek replied:
I reordered it as last

panstromek commented:
I see..  I thought the same thing but I thought that it will be cheap because it is cached. I will reorder it ;) thx

panstromek commented:
@piotrtomiak I addressed your points (hopefully) 😉

After some investigation I decided to make this intention only available for Vue components for now, because Vue actually generates pretty complicated code for some native HTML elements. In those cases it doesn't really seem useful to expand it and I also didn't want to go into too complicated stuff for the first PR.

About the tests - I can't figure out how to run them. Gradle task `test` doesn't run any. Also all test files in the project miss syntax highlighting, which makes them almost impossible to write. Can you point me to how should I set it up?

btw. what is the convention to resolving review? Does author or reviewer mark it as resolved?

alexander-doroshko commented:
@panstromek thanks for the update! At the moment, Piotr is on vacation, please expect a slight delay with this PR.

panstromek commented:
@alexander-doroshko yup, he told me. No hurry ;) thx

piotrtomiak commented on vuejs/src/org/jetbrains/vuejs/intentions/expandVModel/VueExpandVModelIntention.kt:
```diff
@@ -0,0 +1,68 @@
+
+    private fun isValidVModel(attribute: XmlAttribute): Boolean {
+        val info = VueAttributeNameParser.parse((attribute.name), attribute.parent)
+        return info.kind === DIRECTIVE &&
```
Minor: if you do the cast below, no need to check the kind.

panstromek replied:
My reasoning was that "kind" check is faster than cast, so better check that first.. but I can drop that if you think so..

piotrtomiak replied:
That's an interesting question :) I think that if `info` is already loaded into local variable (i.e. registry) the `instanceof` check will be faster than accessing that variables field and comparing it to another variables field.

piotrtomiak commented on vuejs/src/org/jetbrains/vuejs/intentions/expandVModel/VueExpandVModelIntention.kt:
```diff
@@ -0,0 +1,68 @@
+    override fun isAvailable(project: Project, editor: Editor?, element: PsiElement): Boolean {
+        return element.node.elementType == XmlElementType.XML_NAME &&
+                element.parent?.node?.elementType == XmlElementType.XML_ATTRIBUTE &&
+                element.parent.isValid &&
```
I think you should add a check for the component descriptor here as well.

panstromek replied:
Is it really needed? I don't use it anywhere.. How should I check it, in case?

piotrtomiak replied:
You look for it in the `invoke` logic:
```
 val componentDescriptor = componentTag.descriptor as? VueElementDescriptor ?: return
```
And if the descriptor is missing nothing happens, such behavior might be confusing for users.

piotrtomiak commented on vuejs/src/org/jetbrains/vuejs/intentions/expandVModel/VueExpandVModelIntention.kt:
```diff
@@ -0,0 +1,68 @@
+
+    override fun invoke(project: Project, editor: Editor?, psiElement: PsiElement) {
+        editor ?: return
+        val parent: PsiElement = psiElement.parent
```
Minor: maybe one variable for attribute: `val attribute = psiElement.parent as? XmlAttribute ?: return` ?

panstromek replied:
will do

piotrtomiak commented:
Closing the PR due to lack of activity — feel free to create a new PR with the suggested fixes.

panstromek commented:
Yes, sorry. After the last round of reviews, the build broke again, and I didn't have enough enthusiasm to spend another half a day trying to fix it.

piotrtomiak commented:
@panstromek Understood. I've made a lot of improvements to the gradle setup. I'll see if I can rebase this change onto master and merge it with the proposed changes. Sounds ok for you?

panstromek commented:
Ok, that'd be cool. I guess it's not very much needed feeture, though, so probably not worth spending too much time on it if it's broken

piotrtomiak commented:
I've applied your patch, made some very minor changes and merged it. Good work, thanks :)

panstromek commented:
Nice ;) thanks

panstromek commented:
Can you point me to the commit? I can't find it..

piotrtomiak commented:
@panstromek - it takes some time for the commit to be published to GitHub, as this repository is a mirror of a directory in our internal repository.
